README.md
# Millfork_Mega65
repo for Mega65 &amp; Millfork &amp; Shallan's compo base


b.bat and it's variants b_16color.bat etc. are used to build the demos

r.bat and it's variants r_16color.bat etc. are used to launch the demo on mega65 or nexys.

you may need to adjust to suit

#

mega.dmatest.mfk is hello world 

# 16 colors

![16 color simple](/shots/m65_16.png?raw=true)

mega.16color.mfk uses DMA and sets up 16 color mode 

# RRB 

![RRB example](/shots/m65_rrb.png?raw=true)

mega.rrb.mfk DMA 16 color mode and working RRB objects

# IRQ

![IRQ](/shots/m65_irq.png?raw=true)


# tools

tools\bmp2m65 added as a stand alone tool to convert bmp to m65 format tiles and rrb strips

-b 8000 

sets base tile address to $8000 


-o basename 

output is 

basename.chrs 

basename.clut 

basename.map 

basename.atr


-m 

output a traditional tilemap 16x8 pixels 

-r 

output an RRB object, which is split into 16 pixel * height strips

plus 1 blank tile above and 1 below the image for RRB Y offset 

-s

 output sprite strips from image 


# extras

there is an older tilemizer.py script that is no longer supported. but left for reference

you will need python 3 and pillow installed for the image script to work. 











millfork-0.3.28/CHANGELOG.md
# Change log

## 0.3.28 (2021-05-24)

* Officially deprecated decimal operators with apostrophes.

* Added optimization hints.

* Added `utf32be`, `utf32le`, `cp1253`, `cp1254`, `cp1257`, `geos_de` encodings.

* Allowed for underscores in numeric literals for readability purposes, similar to several other programming languages. 

* Added a warning for comparisons between bytes and pointers (#110).

* Fixed escape sequences in many encodings.

* Fixed and documented absolute module imports (#106)
  
* Fixed and optimized sign extension.
  
* Fixed optimizations involving unused labels.
  
* Fixed pointer types to type aliases.

* Fixed parsing of Intel hex literals of the form `0BH`, `0B0H` etc. 

* 6809: Fixed flow analysis in optimization.

* Optimization of certain bitmask operations.

* Parsing optimizations.

## 0.3.26 (2021-03-01)

* Array fields in structs.

* Various Apple II-related improvements, including ProDOS support (thanks to @retrac0).

* Segment-related constants now match their equivalents from the platform definition. Missing constants have been defined.

* Constants with heap start and segment start are now generated properly.

* Signed multiplication support for `sbyte` and `signed16`.

* Heavily experimental `typeof` builtin.

* Self-modifying assembly code is now supported (#101).

* Successful compilation now prints result program size.

* Warning about data not being included in the output file.

* Warnings can now be enabled and disabled individually.

* Imported modules are now identified by their full relative path, not just the token used in the `import` statement (#89).

* 6502: Fixed sbyte to word promotions in certain contexts.

* 8080: Fixed compilation of sign extension of `sbyte` values into the BC register pair.

* Fixed negative constant folding.

* Fixed optimizations around macro invocations.

* 6502: Fixed code deduplication in presence of trampolined functions.

* Optimized word shifts for between 7 and 12 bits.

* Allowed new lines after `=`.

* Various optimization improvements.

* Improved some error messages (thanks to @agg23).

* Other fixes and improvements.

## 0.3.24 (2020-12-02)

* Preliminary support for TRS-80 Model 1 and 3 running TRS-DOS.

* Preliminary support for Robotron Z1013.

* Allowed defining entry points other than the start of the segment for Atari, ZX Spectrum, CoCo, Z1013 and TRS-80. (#78)

* Allowed the `:` operator in const-pure functions.

* Added `pointer.interrupt` and `pointer.kernal_interrupt` types.

* Implemented `readline` and `readword` for VIC-20.

* `init_rand_seed` uses the POKEY on Atari.

* Useless labels are no longer emitted into the label file.

* VIC-20: added `readline` and `readword`.

* Atari: use POKEY for randomness source.

* New output format elements: ASCII string, program name.

* Fix: Pointers to functions with parameters (#86)/

* Fix: more instances of memset loops should be detected and optimized (#59).

* Fix: things mentioned in the segment layout should not be deleted even if unused.

* Fix: `endaddr+N` output format.

* 65816: some code generation fixes.

* 8080: word negation now works.

* Various optimization improvements.

* Various other fixes.

* Improved some error messages.

* Even more new Atari examples (thanks to @zbyti).

* Build process slightly changed.

## 0.3.22 (2020-09-15)

* Added local labels in assembly.

* Added alternate decimal operators (with `$` instead of `'`).

* **Potentially breaking change!** Identifiers no longer can end with `$`.

* Added `z80next` as an alternate name for the ZX Spectrum Next's processor (#55).

* Added encodings: `brascii`, `macroman`, `dmcs`, `lics`.

* Improved some error messages.

* Fix: interrupt functions written in assembly no longer have the default prologue (#62).

* Fixed the `a8_os` module (#58).

* Fixed evaluation of division of large constants.

* Fix: Structure alignment is now respected for substructures.

* X16: Fixed the address of `vera_dc_hscale_hstop` register (#54) (thanks to @Kobrasadetin).

* Fixed evaluation of more complex boolean expressions (#56).

* Fixed accesses to volatile variables.

* 8080/Z80: Optimization improvements. 

* 6809: Optimization improvements.

* Various Atari improvements (#60, #63, #65) (thanks to @zbyti).

* New Atari examples (thanks to @zbyti).

## 0.3.18 (2020-04-08)

* Support for Motorola 6809 (complete, but still experimental).

* Preliminary support for Tandy Color Computer running RS-DOS.

* Preliminary support for 16K cartridges for Tandy Color Computer.

* Added support for modifying large variables via pointers.

* Added the ability to declare structure alignment.

* `for` loops over arrays.

* Allowed defining custom text encodings.
**Potentially breaking change!**
There are no built-in encodings now, the include path needs to contain the necessary encodings.

* Fixed encodings: 
`apple2`, `atasciiscr`, `iso_de`, `iso_no`, `iso_se`, 
`koi7n2`, `msx_jp`, 
`oldpet`, `origpet`,  `petscii`, `petsciijp`, `petscr`, `petscrjp`, 
`zx80`.

* Added encodings:
`apple2c`, `apple2e`, `apple2gs`,
`coco`, `cocoscr`,
`cpc_da`, `cpc_en`, `cpc_es`, `cpc_fr`,
`cp437`, `cp850`, `cp851`, `cp852`, `cp855`, `cp858`, `cp866`, 
`cp1250`, `cp1251`, `cp1252`,
`ebcdic`,
`galaksija`,
`iso8859_1`, `iso8859_2`, `iso8859_3`, `iso8859_4`, `iso8859_5`,
`iso8859_7`, `iso8859_9`, `iso8859_10`, `iso8859_13`, `iso8859_14`, `iso8859_16`,
`kamenicky`,
`koi8e`, `koi8f`, `koi8r`, `koi8ru`, `koi8t`, `koi8u`,
`mazovia`, `pcw`,
`pokemon1en`, `pokemon1es`, `pokemon1fr`, `pokemon1jp`.

* Added `ENCODING_NOLOWER` preprocessor feature.

* Fixed raw views of typed pointers.

* Fixed dead code elimination (#51). 

* **Potentially breaking change!** Changed default encoding for CPC to `cpc_en`.


tools/dmt.c
/**
 * Copyright (c) 2013, rxi
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the MIT license. See LICENSE for details.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>

//#define DMT_STACK_TRACE
#include "dmt.h"

#ifdef DMT_STACK_TRACE
#include <execinfo.h>
#ifndef DMT_STACK_TRACE_MAX
#define DMT_STACK_TRACE_MAX 32
#endif
#endif


typedef struct dmt_node_t {
  struct dmt_node_t *prev, *next;
  const char *file;
  size_t line;
  size_t size;
#ifdef DMT_STACK_TRACE
  void  *stacktrace[DMT_STACK_TRACE_MAX];
  size_t stacktrace_sz;
#endif
} dmt_node_t;


dmt_node_t *dmt_head;



int _dmt_has_node(dmt_node_t *n) {
  dmt_node_t *node = dmt_head;
  while (node != NULL) {
    if (node == n) return 1;
    node = node->next;
  }
  return 0;
}



void _dmt_abort(void) {
#ifdef DMT_STACK_TRACE
  void *array[DMT_STACK_TRACE_MAX];
  size_t sz = backtrace(array, DMT_STACK_TRACE_MAX);
  backtrace_symbols_fd(array, sz, fileno(stderr));
#endif
  abort();
}

uint64_t total_allocated = 0;

void   *_dmt_strdup(const char *s, const char*file, unsigned line)
{
unsigned len = (unsigned)strlen(s)+1;
unsigned char *bff=_dmt_alloc(len,0,file,line);
	memcpy(bff,s,len);
	return bff;
}

void *_dmt_alloc(size_t sz, int zeroset, const char *file, unsigned line) {
  dmt_node_t *node = NULL;
  
  if (zeroset) {
    node = calloc(sizeof(*node) + sz, 1);
  } else {
    node = malloc(sizeof(*node) + sz);
    if (node != NULL) {
      memset(node, 0, sizeof(*node));
    }
  }

  if (node == NULL) {
#ifdef DMT_ABORT_NULL
    fprintf(stderr, "Couldn't allocate: %s, line %u\n", file, line);
    _dmt_abort();
#else
    return NULL;
#endif
  }

  node->line = line;
  node->file = file;
  node->size = sz;

#ifdef DMT_STACK_TRACE
  node->stacktrace_sz = backtrace(node->stacktrace, DMT_STACK_TRACE_MAX);
#endif

	total_allocated+=sz;

  if (dmt_head) {
    dmt_head->prev = node;
    node->next = dmt_head;
  }
  dmt_head = node;

  return (char*)node + sizeof(*node);
}



void *_dmt_realloc(void *ptr, size_t sz, const char *file, unsigned line) {
  dmt_node_t *node = (dmt_node_t*)((char*)ptr - sizeof(*node));
  dmt_node_t *old_node = node;

  if (ptr == NULL) return _dmt_alloc(sz, 0, file, line);

#ifndef DMT_UNSAFE
  if (!_dmt_has_node(node)) {
    fprintf(stderr, "Bad realloc: %p %s, line %u\n", ptr, file, line);
    _dmt_abort();
  }
#endif

  node = realloc(node, sizeof(*node) + sz);

  if (node == NULL) {
#ifdef DMT_ABORT_NULL
    fprintf(stderr, "Couldn't reallocate: %s, line %u\n", file, line);
    _dmt_abort();
#else
    return NULL;
#endif
  }

  node->size = sz;
  if (dmt_head == old_node) dmt_head = node;
  if (node->prev) node->prev->next = node;
  if (node->next) node->next->prev = node;

  return (char*)node + sizeof(*node);
}



void _dmt_free(void *ptr, const char *file, unsigned line) {
  dmt_node_t *node = (dmt_node_t*)((char*)ptr - sizeof(*node));

  if (ptr == NULL) return;
#ifndef DMT_UNSAFE
  if (!_dmt_has_node(node)) {
    fprintf(stderr, "Bad free: %p %s, line %u\n", ptr, file, line);
    _dmt_abort();
  }
#endif

//	fprintf(stderr,"free %x %s,line %u %u\n",ptr,file,line,node->size);

  if (node == dmt_head) dmt_head = node->next;
  if (node->prev) node->prev->next = node->next;
  if (node->next) node->next->prev = node->prev;

  free(node);
}



void dmt_dump(FILE *fp) {
  dmt_node_t *node = dmt_head;
  size_t total = 0;

  if (!fp) fp = stdout;

  while (node != NULL) {
    fprintf(fp, "Unfreed: %p %s, line %lu (%lu bytes)\n", 
            (char*)node + sizeof(*node), node->file,
            (unsigned long)node->line, (unsigned long)node->size);

#ifdef DMT_STACK_TRACE
    backtrace_symbols_fd(node->stacktrace, node->stacktrace_sz, fileno(fp));
    fprintf(fp, "\n");
#endif

    total += node->size;
    node = node->next;
  }

  fprintf(fp, "Total unfreed: %lu bytes\n", (unsigned long)total);
	fprintf(fp, "Total allocated: %f kb\n",(float)total_allocated/1024.0f);
}



size_t _dmt_size(void *ptr, const char* file, unsigned line) {
  dmt_node_t *node = (dmt_node_t*)((char*)ptr - sizeof(*node));

#ifndef DMT_UNSAFE
  if (!_dmt_has_node(node)) {
    fprintf(stderr, "Bad pointer: %p %s, line %u\n", ptr, file, line);
    _dmt_abort();
  }
#endif

  return node->size;
}



size_t dmt_usage(void) {
  dmt_node_t *node = dmt_head;
  size_t total = 0;

  while (node != NULL) {
    total += node->size;
    node = node->next;
  } 

  return total;
}



int dmt_has(void *ptr) {
  dmt_node_t *node = (dmt_node_t*)((char*)ptr - sizeof(*node));
  return _dmt_has_node(node);
}


tools/dmt.h
/**
 * Copyright (c) 2013, rxi
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the MIT license. See LICENSE for details.
 */

#ifndef DMT_H_
#define DMT_H_

#include <stdio.h>
#ifdef	__cplusplus
extern "C" {
#endif

void _dmt_free(void *ptr, const char *file, unsigned line);
void *_dmt_strdup(const char *s, const char*file, unsigned line);

#ifdef	__cplusplus
}
#endif


#define dmt_malloc(sz)        _dmt_alloc(sz, 0, __FILE__, __LINE__)
#define dmt_calloc(num, sz)   _dmt_alloc((num) * (sz), 1, __FILE__, __LINE__)
#define dmt_realloc(ptr, sz)  _dmt_realloc(ptr, sz, __FILE__, __LINE__)
#define dmt_free(ptr)         _dmt_free(ptr, __FILE__, __LINE__)
#define dmt_size(ptr)         _dmt_size(ptr, __FILE__, __LINE__)
#define dmt_strdup(sz)        _dmt_strdup(sz, __FILE__, __LINE__)

void   *_dmt_strdup(const char *, const char*, unsigned);
void   *_dmt_alloc(size_t, int, const char*, unsigned);
void   *_dmt_realloc(void*, size_t, const char*, unsigned);
void    _dmt_free(void*, const char*, unsigned);
size_t  _dmt_size(void*, const char*, unsigned);

void    dmt_dump(FILE*);
size_t  dmt_usage(void);
int     dmt_has(void *ptr);

#endif


tools/tilemizer.py
import sys
import os
from PIL import Image, ImageSequence, ImageDraw
from itertools import groupby
import subprocess

def StoreByte(file,value):
	value&=0xff
	file.write(value.to_bytes(1,byteorder='little'))

def StoreByteSwapped(file,value):
	value&=0xff
	a = (value>>4) | ((value&0xf)<<4)
	file.write(a.to_bytes(1,byteorder='little'))


def StoreWord(file,value):
	global binaryData
	value&=0xffff
	file.write(value.to_bytes(2,byteorder='little'))


class Settings:
	def __init__(self):
		self.tile_width = 16
		self.tile_height = 8
		self.meta_width = -1 	# tiles
		self.meta_height = -1	# tiles
		self.outaddress = 0x8000
		self.offset = 0x200
		self.output = "4bpp"
		self.mask = 0xf 
		self.palette = 0
		self.outname = "output"
		self.check_repeats = True
# split into tiles

def Spriteize(name):
	global binaryData,attrib,index,settings
	image = Image.open(name)
	(w, h) = image.size
	tiles = []
	(mw,mh) = (w/settings.tile_width,h/settings.tile_height)

	for ty in range(0,h,settings.tile_height):
		for tx in range(0,w,settings.tile_width):
			blank = []
			index = -1
			palette_index = 0
			# check non flipped
			for py in range(0,settings.tile_height):
				for px in range(0,settings.tile_width):
					p = image.getpixel((tx+px,ty+py))
					blank.append(p & settings.mask)
					# pull palette index if not currently 0 
					if (p&0xf0)>0:
						palette_index = p>>4

			#	check if we have this tile
#			if blank in tiles:
#				index = tiles.index(blank)
#			if (index==-1):
			index = len(tiles)
			tiles.append(blank)

	binaryData = open(settings.outname + ".spr", 'wb')
	for t in tiles:
		for py in range(0,settings.tile_height):
			if settings.output == "8bpp":
				for px in range(0,settings.tile_width):
					b = t[px+(py*settings.tile_width)]
					StoreByte(binaryData,b)	
			if settings.output == "4bpp":
				for px in range(0,settings.tile_width,2):
					a = t[px+(py*settings.tile_width)]
					b = t[(1+px)+(py*settings.tile_width)]
					StoreByte(binaryData,(a<<4) | b)	
	binaryData.close()

	# save palette data
	# nybble swapped and in seperate channels 
	binaryData = open(settings.outname + ".clut", 'wb')
	rgb = image.getpalette()
	for r in range(0,256):
		StoreByteSwapped(binaryData,rgb[(r*3)+0])
	for r in range(0,256):
		StoreByteSwapped(binaryData,rgb[(r*3)+1])
	for r in range(0,256):
		StoreByteSwapped(binaryData,rgb[(r*3)+2])
	binaryData.close()	


def Objectize(name):
	global binaryData,attrib,index,settings,tiles

	# generate long strips to decompress 

	image = Image.open(name)
	(w, h) = image.size
	tiles = []
#	map = []
#	attribs = []	

	(mw,mh) = (w/settings.tile_width,h/settings.tile_height)

	if mw.is_integer()==False:
		print("map width isn't evenly divisible by " + str(settings.tile_width))
	if mh.is_integer()==False:
		print("map height isn't evenly divisible by " + str(settings.tile_height))

	if settings.meta_width==-1:
		settings.meta_width = int(mw) 
		settings.meta_height = int(mh)

	print("map " + str(mw) + "," + str(mh) + " meta " + str(settings.meta_width) + "," + str(settings.meta_height))

	# store the width and height of the map
	binaryData = open(settings.outname + ".info", 'wb')
	StoreWord(binaryData,int(mw))
	StoreWord(binaryData,int(mh))
	binaryData.close()

	mapData = open(settings.outname + ".map","wb")
	attrData = open(settings.outname + ".atr","wb")
	id = 1
	binaryData = bytearray()

	offset = 0 
	offsets = []
	data = []
	lines = []

	uniques = []
	

	for frame in ImageSequence.Iterator(image):		
		buffer = []
		for x in range(0,w):
			for y in range(0,h):
				p = frame.getpixel((x,y))
				buffer.append(p)
		if buffer not in uniques:
			uniques.append(buffer)			
			filename = settings.outname + "_" + hex(id)[2:] + ".str"
			mapData = open("bin/raw.bin","wb")

			print(filename)
			id=id+1
#			rawPixels = open("bin/pic" + "_" + hex(id)[2:] + ".bin","wb")
#			for x in range(0,w):
#				for y in range(0,h):
#					p = frame.getpixel((x,y))
#					StoreByte(rawPixels,p)
#			rawPixels.close()

			for tx in range(0,w,settings.tile_width):
				for y in range(0,h,settings.tile_height):
					for py in range(0,settings.tile_height):
						for px in range(0,settings.tile_width,2):
							p = frame.getpixel((tx+px,py+y)) & settings.mask
							q = frame.getpixel((tx+px+1,py+y)) & settings.mask

							StoreByte(mapData,q<<4 | p)
			mapData.close()

	#		result = subprocess.call("../tools/apultra.exe " + "bin/raw.bin" + " " + filename , shell=False)	
	#		result = subprocess.call("../tools/rle.exe " + "bin/raw.bin" + " " + filename , shell=False)	
	#		result = subprocess.call("../tools/lzsa.exe --prefer-speed " + "bin/raw.bin" + " " + filename , shell=False)	
	#		result = subprocess.call("../tools/c64f.exe " + "bin/raw.bin" + " " + filename , shell=False)	
	#		result = subprocess.call("../tools/zpack.exe " + "bin/raw.bin" + " " + filename , shell=False)	


#			result = subprocess.call("python ../tools/tc_encode.py -r " + "bin/raw.bin" + " " + filename , shell=False)	

#			result = subprocess.call("../tools/packbits.exe " + "bin/raw.bin" + " " + filename , shell=False)	
#			result = subprocess.call("../tools/lzg.exe -9 " + "bin/raw.bin" + " " + filename , shell=False)	
#			result = subprocess.call("../tools/zx0.exe " + "bin/raw.bin" + " " + filename , shell=False)	


			result = subprocess.call("python ../tools/rle.py " + "bin/raw.bin" + " " + filename , shell=False)	
#			result = subprocess.call("../tools/c64f.exe " + "bin/raw.bin" + " " + filename , shell=False)	

			offsets.append(os.stat(filename).st_size)
	#		file = open("bin/raw.bin", "rb")
			file = open(filename, "rb")
			buffer = file.read()


			data.append(buffer)
			file.close()
			result = subprocess.call("rm " + filename , shell=False)	



	mapData = open(settings.outname + ".pak","wb")
#
	baseoffset = 4+(len(offsets)*2)

	binaryData.append(int(mw*2)&0xff)
	binaryData.append(int(mh)&0xff)
	binaryData.append(int(len(offsets)))
	binaryData.append(int(len(offsets))>>8)

	datasize = int((mw * 6) * mh)	# 2 char buffers + 1 attrib buffer   

	base = baseoffset 

	# n frames
#	StoreWord(mapData,len(names))
	base = baseoffset
	for x in offsets:
		binaryData.append(base&0xff)
		binaryData.append(base>>8)
		base=base+x 

	for x in data:
		binaryData+=x

	mapData.write(binaryData)
	mapData.close()
#	print(offsets)
#	print(hex(int(datasize)))

def Tilemize(name):
	global binaryData,attrib,index,settings,tiles

	image = Image.open(name)
	(w, h) = image.size
	tiles = []
#	map = []
#	attribs = []	

	(mw,mh) = (w/settings.tile_width,h/settings.tile_height)


	if mw.is_integer()==False:
		print("map width isn't evenly divisible by " + str(settings.tile_width))
	if mh.is_integer()==False:
		print("map height isn't evenly divisible by " + str(settings.tile_height))

	if settings.meta_width==-1:
		settings.meta_width = int(mw) 
		settings.meta_height = int(mh)

	print("map " + str(mw) + "," + str(mh) + " meta " + str(settings.meta_width) + "," + str(settings.meta_height))

	# store the width and height of the map
	binaryData = open(settings.outname + ".info", 'wb')
	StoreWord(binaryData,int(mw))
	StoreWord(binaryData,int(mh))
	binaryData.close()

	mapData = open(settings.outname + ".map","wb")
	attrData = open(settings.outname + ".atr","wb")
	id = 1
	for frame in ImageSequence.Iterator(image):		
		# chop into tiles
#		print("frame " + str(id))
		id=id+1
		for ty in range(0,h,settings.tile_height*settings.meta_height):
			for tx in range(0,w,settings.tile_width*settings.meta_width):
				# for each type of char 
#				print(str(tx) + "," + (str(ty)))
				map = [0] * (settings.meta_width * settings.meta_height)
				attribs = [0] * (settings.meta_width * settings.meta_height)
				for mtx in range(0,settings.meta_width):
					for mty in range(0,settings.meta_height):
						global index
						blank = []
						blank_x = []
						blank_y = []
						blank_xy = []

						index = -1
						# default attribs for 4bpp
						if settings.output == "4bpp":
							attrib = 0x0f08
						else:
							attrib = 0x0f00

	#					if settings.palette!=0:
	#						attrib= attrib|0x6000

						palette_index = 0
						unique = False
						# check non flipped
						bx = tx + (mtx * settings.tile_width)
						by = ty + (mty * settings.tile_height)
						for py in range(0,settings.tile_height):
							for px in range(0,settings.tile_width):
								p = frame.getpixel((bx+px,by+py))
	#							if (p&0xf)==0xf:
	#								unique = True
	#								p=p&0xf0

								blank.append(p & settings.mask)
								# pull palette index if not currently 0 
								if (p&0xf0)>0:
									palette_index = p>>4
						# set palette attribs
						attrib|=palette_index<<12
						if settings.check_repeats==True:
							#	check if we have this tile
							if blank in tiles:
								index = tiles.index(blank)
							# check x flipped
							if index == -1:
								for py in range(0,settings.tile_height):
									for px in range(0,settings.tile_width):
										p = frame.getpixel((bx+(settings.tile_width-1-px),by+py)) & settings.mask
										blank_x.append(p)
								#	lets see if we have this tile flipped in X
								if blank_x in tiles:
									index = tiles.index(blank_x)
									#	set flipped attributes
									attrib|=1<<6
			
			
							# check y flipped
							if index == -1:
								for py in range(0,settings.tile_height):
									for px in range(0,settings.tile_width):
										p = frame.getpixel((bx+px,(by+(settings.tile_height-1-py)))) & settings.mask
										blank_y.append(p)
								#	lets see if we have this tile flipped in Y
								if blank_y in tiles:
									index = tiles.index(blank_y)
									#	set flipped attributes
									attrib|=1<<7
			
							# check xy flipped
							if index == -1:
								for py in range(0,settings.tile_height):
									for px in range(0,settings.tile_width):
										p = frame.getpixel((bx+(settings.tile_width-1-px),(by+(settings.tile_height-1-py)))) & settings.mask
										blank_xy.append(p)
								#	lets see if we have this tile flipped in X&Y
								if blank_xy in tiles:
									index = tiles.index(blank_xy)
									#	set flipped attributes
									attrib|=1<<6
									attrib|=1<<7
			
	#					#	if we've never seen this one , add it 
						if (index==-1) or (unique==True):
							index = len(tiles)
							tiles.append(blank)
	#
	#
						offset = mtx + (mty*settings.meta_width)
						map[offset] = index + settings.offset
						attribs[offset] = attrib
						#	fill the map data with data we got above

#				s = ""
				for q in map:
#					s = s + hex(q) + ","
					StoreWord(mapData,q)
#				print(s)
				for q in attribs:
					StoreWord(attrData,q)

#				map.append(0x0140)
#				map.append(0x0000)
#				attribs.append(0x0090)
#				attribs.append(0x0000)

		#	save map / screen data

#			map.append(0x0140)
#			attribs.append(0x0010)
#			map.append(0x0000)
#			attribs.append(0x0000)


	print(str(len(tiles)) + " tiles found")

	#	save 4bpp char data ( or 8 bpp )
	binaryData = open(settings.outname + ".chrs", 'wb')
	for t in tiles:
		for py in range(0,settings.tile_height):
			s = "\t"
#									map[offset] = index + settings.offset
			if settings.output == "8bpp":
				for px in range(0,settings.tile_width):
					b = t[px+(py*settings.tile_width)]
					StoreByte(binaryData,b)	
			if settings.output == "4bpp":
				for px in range(0,settings.tile_width,2):
					a = t[px+(py*settings.tile_width)]
					b = t[(1+px)+(py*settings.tile_width)]
					StoreByte(binaryData,(b<<4) | a)
					if a==0:
						s=s + "."
					else:
						s = s + hex(a)[2:].upper()
					if b==0:
						s=s + "."
					else:
						s = s + hex(b)[2:].upper()
#			print(s)

	binaryData.close()

	mapData.close()

	#	save colors / attributes
	attrData.close()

	# save palette data
	# nybble swapped and in seperate channels 
	binaryData = open(settings.outname + ".clut", 'wb')
	rgb = image.getpalette()
	for r in range(0,256):
		StoreByteSwapped(binaryData,rgb[(r*3)+0])
	for r in range(0,256):
		StoreByteSwapped(binaryData,rgb[(r*3)+1])
	for r in range(0,256):
		StoreByteSwapped(binaryData,rgb[(r*3)+2])

	binaryData.close()	

def main():
	global binaryData,attrib,index,settings
	settings = Settings()

	for a in range(1,len(sys.argv)):
		arg = sys.argv[a]
		print(arg)

		if "-b4" in arg:
			settings.tile_width = 16
			settings.output = "4bpp"
			settings.mask = 0xf
		elif "-b8" in arg:
			settings.tile_width = 8
			settings.output = "8bpp"
			settings.mask = 0xff
		elif arg.startswith("-m="):
			settings.outaddress = int(arg[3:],16)
			settings.offset = settings.outaddress//64
		elif arg.startswith("-o="):
			settings.outname = arg[3:]
		elif arg.startswith("-t="):
			name = arg[3:]
			settings.check_repeats = True
			Tilemize(name)
		elif arg.startswith("-tn="):
			name = arg[4:]
			settings.check_repeats = False
			Tilemize(name)
		elif arg.startswith("-p="):
			name = arg[3:]
			Objectize(name)
		elif arg.startswith("-s="):
			settings.tile_height = 128
			name = arg[3:]
			Spriteize(name)
		elif arg.startswith("-mt="):
			settings.meta_width = 8 
			settings.meta_height = 32
			settings.tile_height = 8
			settings.check_repeats = False
			name = arg[4:]
			Tilemize(name)



if __name__ == '__main__':
   main()


